Plugins/Cuda/src/Utilities/MemoryManager.cu

```diff
@@ -0,0 +1,144 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// CUDA plugin include(s).
+#include "Acts/Plugins/Cuda/Utilities/MemoryManager.hpp"
+#include "ErrorCheck.cuh"
+
+// CUDA include(s).
+#include <cuda_runtime.h>
+
+// System include(s).
+#include <cmath>
+#include <stdexcept>
+
+namespace Acts {
+namespace Cuda {
+
+MemoryManager::~MemoryManager() {
+  // Free all the allocated memory.
+  for (DeviceMemory& mem : m_memory) {
+    if (mem.m_ptr == nullptr) {
+      continue;
+    }
+    ACTS_CUDA_ERROR_CHECK(cudaFree(mem.m_ptr));
+  }
+}
+
+MemoryManager& MemoryManager::instance() {
+  static MemoryManager mm;
+  return mm;
+}
```

Comment:
I assume there's no point in having multiple memory managers?

Reply:
The reason I chose to use a singleton design is to not have to pass a memory manager object explicitly to every single `Acts::Cuda::device_array<...>(...)` object that the code creates. And yes, the idea would be that we should centralise all memory allocations/de-allocations. So the singleton design would seem to fit the use case quite nicely.

But I definitely don't claim that this would be some great implementation for a CUDA memory manager. I would like to make this code a lot more advanced later on. Using code inspired by (stolen from... :stuck_out_tongue:) from [Allen](https://arxiv.org/abs/1912.09161) for instance. :wink:

Reply:
Ok. I don't think the singleton pattern is problematic here, for now at least.